README.md
<p align="center">
    <img src="images/logo.jpg"
        alt="StayAwake logo" />
</p>

[![CI](https://github.com/krystianhub/StayAwake/actions/workflows/Build.yml/badge.svg?branch=master)](https://github.com/krystianhub/StayAwake/actions/workflows/Build.yml)
[![codecov](https://codecov.io/gh/krystianhub/StayAwake/branch/master/graph/badge.svg?token=S84RWLU7YA)](https://codecov.io/gh/krystianhub/StayAwake)
[![crates.io](https://img.shields.io/crates/v/stayawake.svg)](https://crates.io/crates/stayawake)
[![license](https://img.shields.io/badge/license-MIT-blue.svg)](https://github.com/krystianhub/StayAwake/blob/master/LICENSE)

# StayAwake

## Introduction

StayAwake is a simple program for Windows & MacOS to keep your system awake without affecting your workflow.

It is inspired by the [stay-awake](https://pypi.org/project/stay-awake/) package for Python.

The program is only triggered when you don't do any mouse movements and it is completely headless, intended to be used as a command line tool.

### How does it work?

If in a span of **15** seconds you don't move your mouse, StayAwake will automatically move your mouse for about **100** to **150** pixels randomly. There won't be any mouse displacement! If you are working, this will do absolutely nothing!

All settings like the time interval or the range of random mouse movement are fully customizable via environment variables or the **[.env](.env)** file.

## Installation

You can download the [latest version](https://github.com/krystianhub/StayAwake/releases/latest) from Github Releases.

Alternatively, you can install it via **cargo** command:

```bash
cargo install stayawake
```

## Configuration

The application ships with a default configuration, but it can be overridden by creating a **[.env](.env)** file in the same location as the binary itself.

There are 6 available configuration properties:

```properties
RUST_LOG=INFO # logging level
STAYAWAKE_INTERVAL=15 # in seconds
JUMP_BY_PIXEL_MIN=100 # in pixels
JUMP_BY_PIXEL_MAX=150 # in pixels
INIT_POINT=0x0 # in pixels, starting point of the operating window area (usually it's a top-left corner of the screen; for example, 0x0)
WORKING_AREA=1024x768 # in pixels, operating window area (usually it's a display resolution; for example, 1024x768)
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.2.4] - 2023-03-19

### Changed

- Updated dependencies
- Updated CI/CD pipelines

## [0.2.3] - 2022-10-20

### Changed

- Updated dependencies

## [0.2.2] - 2022-06-07

### Changed

- Updated dependencies

## [0.2.1] - 2022-04-05

### Added

- (CI/CD) Codecov.io integration
- Improved unit test coverage

## [0.2.0] - 2022-03-30

### Added

- Two new configuration properties called `INIT_POINT` and `WORKING_AREA`
- Basic unit test coverage

### Changed

- Renamed `OFFSET_PIXEL_MIN` to `JUMP_BY_PIXEL_MIN`
- Renamed `OFFSET_PIXEL_MAX` to `JUMP_BY_PIXEL_MAX`
- Updated dependencies

## [0.1.2] - 2022-02-22

### Changed

- Upgraded to 2021 edition of Rust

## [0.1.1] - 2021-07-01

### Added

- Support for inhibiting Power Management (Windows & MacOS)
- [CHANGELOG.md](CHANGELOG.md) file
- [LICENSE](LICENSE) file

## [0.1.0] - 2021-06-10

### Added

- First official release

[Unreleased]: https://github.com/krystianhub/StayAwake/compare/v0.2.4...HEAD
[0.2.4]: https://github.com/krystianhub/StayAwake/compare/v0.2.3...v0.2.4
[0.2.3]: https://github.com/krystianhub/StayAwake/compare/v0.2.2...v0.2.3
[0.2.2]: https://github.com/krystianhub/StayAwake/compare/v0.2.1...v0.2.2
[0.2.1]: https://github.com/krystianhub/StayAwake/compare/v0.2.0...v0.2.1
[0.2.0]: https://github.com/krystianhub/StayAwake/compare/v0.1.2...v0.2.0
[0.1.2]: https://github.com/krystianhub/StayAwake/compare/v0.1.1...v0.1.2
[0.1.1]: https://github.com/krystianhub/StayAwake/compare/v0.1.0...v0.1.1
[0.1.0]: https://github.com/krystianhub/StayAwake/releases/tag/v0.1.0


src/config.rs
use crate::{
    config::ConfigError::InvalidProperty,
    models::{InitPoint, WorkingArea},
};
use serde::Deserialize;
use serde_with::{serde_as, DurationSeconds};
use std::time::Duration;
use thiserror::Error;

/// Provides default value for stayawake_interval if STAYAWAKE_INTERVAL env var is not set
fn default_stayawake_interval() -> Duration {
    Duration::from_secs(15)
}

/// Provides default value for jump_by_pixel_min if JUMP_BY_PIXEL_MIN env var is not set
fn default_jump_by_pixel_min() -> usize {
    100
}

/// Provides default value for jump_by_pixel_max if JUMP_BY_PIXEL_MAX env var is not set
fn default_jump_by_pixel_max() -> usize {
    150
}

/// Provides default value for init_point if INIT_POINT env var is not set
fn default_init_point() -> InitPoint {
    InitPoint { x: 0, y: 0 }
}

/// Provides default value for working_area if WORKING_AREA env var is not set
fn default_working_area() -> WorkingArea {
    WorkingArea {
        width: 1024,
        height: 768,
    }
}

#[derive(Error, Debug)]
pub enum ConfigError {
    #[error("Configuration property `{property}` is invalid: {message}")]
    InvalidProperty {
        property: &'static str,
        message: &'static str,
    },
}

/// Configuration struct (.env file)
#[serde_as]
#[derive(Deserialize, Debug)]
pub(crate) struct Config {
    #[serde_as(as = "DurationSeconds<u64>")]
    #[serde(default = "default_stayawake_interval")]
    pub(crate) stayawake_interval: Duration,
    #[serde(default = "default_jump_by_pixel_min")]
    pub(crate) jump_by_pixel_min: usize,
    #[serde(default = "default_jump_by_pixel_max")]
    pub(crate) jump_by_pixel_max: usize,
    #[serde(default = "default_init_point")]
    pub(crate) init_point: InitPoint,
    #[serde(default = "default_working_area")]
    pub(crate) working_area: WorkingArea,
}

impl Config {
    /// Validates if the config is correct
    pub(crate) fn validate(&self) -> Result<(), ConfigError> {
        if self.jump_by_pixel_min == 0 {
            return Err(InvalidProperty {
                property: "jump_by_pixel_min",
                message: "jump_by_pixel_min cannot be equal to zero",
            });
        }

        if self.jump_by_pixel_max == 0 {
            return Err(InvalidProperty {
                property: "jump_by_pixel_max",
                message: "jump_by_pixel_max cannot be equal to zero",
            });
        }

        if self.working_area.width == 0 || self.working_area.height == 0 {
            return Err(InvalidProperty {
                property: "working_area",
                message: "working_area height or/and width cannot be equal to zero",
            });
        }

        // jump by pixel min have to be smaller than max
        if self.jump_by_pixel_min > self.jump_by_pixel_max {
            return Err(InvalidProperty {
                property: "jump_by_pixel_min",
                message: "jump_by_pixel_min cannot be bigger than jump_by_pixel_max",
            });
        }

        // working area size have to be bigger than jump by pixel max
        if self.working_area.width <= self.jump_by_pixel_max
            || self.working_area.height <= self.jump_by_pixel_max
        {
            return Err(InvalidProperty {
                property: "working_area",
                message: "working_area cannot be equal or smaller than jump_by_pixel_max",
            });
        }

        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_config_validate() {
        let config = Config {
            stayawake_interval: default_stayawake_interval(),
            jump_by_pixel_min: default_jump_by_pixel_min(),
            jump_by_pixel_max: default_jump_by_pixel_max(),
            init_point: default_init_point(),
            working_area: default_working_area(),
        };

        assert!(config.validate().is_ok());

        // ----------------

        let config = Config {
            stayawake_interval: default_stayawake_interval(),
            jump_by_pixel_min: 100,
            jump_by_pixel_max: 150,
            init_point: InitPoint { x: 0, y: 0 },
            working_area: WorkingArea {
                width: 50,
                height: 500,
            },
        };

        let result = config.validate();
        assert!(result.is_err());

        let result_err = result.unwrap_err();
        let InvalidProperty { property, message } = result_err;
        assert_eq!(property, "working_area");
        assert_eq!(
            message,
            "working_area cannot be equal or smaller than jump_by_pixel_max"
        );

        // ----------------

        let config = Config {
            stayawake_interval: default_stayawake_interval(),
            jump_by_pixel_min: 100,
            jump_by_pixel_max: 150,
            init_point: InitPoint { x: 0, y: 0 },
            working_area: WorkingArea {
                width: 500,
                height: 50,
            },
        };

        let result = config.validate();
        assert!(result.is_err());

        let result_err = result.unwrap_err();
        let InvalidProperty { property, message } = result_err;
        assert_eq!(property, "working_area");
        assert_eq!(
            message,
            "working_area cannot be equal or smaller than jump_by_pixel_max"
        );

        // ----------------

        let config = Config {
            stayawake_interval: default_stayawake_interval(),
            jump_by_pixel_min: 150,
            jump_by_pixel_max: 150,
            init_point: InitPoint { x: 0, y: 0 },
            working_area: WorkingArea {
                width: 150,
                height: 150,
            },
        };

        let result = config.validate();
        assert!(result.is_err());

        let result_err = result.unwrap_err();
        let InvalidProperty { property, message } = result_err;
        assert_eq!(property, "working_area");
        assert_eq!(
            message,
            "working_area cannot be equal or smaller than jump_by_pixel_max"
        );

        // ----------------

        let config = Config {
            stayawake_interval: default_stayawake_interval(),
            jump_by_pixel_min: 99,
            jump_by_pixel_max: 99,
            init_point: InitPoint { x: 0, y: 0 },
            working_area: WorkingArea {
                width: 100,
                height: 100,
            },
        };

        assert!(config.validate().is_ok());

        // ----------------

        let config = Config {
            stayawake_interval: default_stayawake_interval(),
            jump_by_pixel_min: 99,
            jump_by_pixel_max: 99,
            init_point: InitPoint { x: 50, y: 50 },
            working_area: WorkingArea {
                width: 100,
                height: 100,
            },
        };

        assert!(config.validate().is_ok());

        // ----------------

        let config = Config {
            stayawake_interval: default_stayawake_interval(),
            jump_by_pixel_min: 101,
            jump_by_pixel_max: 100,
            init_point: InitPoint { x: 0, y: 0 },
            working_area: WorkingArea {
                width: 150,
                height: 150,
            },
        };

        let result = config.validate();
        assert!(result.is_err());

        let result_err = result.unwrap_err();
        let InvalidProperty { property, message } = result_err;
        assert_eq!(property, "jump_by_pixel_min");
        assert_eq!(
            message,
            "jump_by_pixel_min cannot be bigger than jump_by_pixel_max"
        );

        // ----------------

        let config = Config {
            stayawake_interval: default_stayawake_interval(),
            jump_by_pixel_min: 0,
            jump_by_pixel_max: 1,
            init_point: InitPoint { x: 0, y: 0 },
            working_area: WorkingArea {
                width: 150,
                height: 150,
            },
        };

        let result = config.validate();
        assert!(result.is_err());

        let result_err = result.unwrap_err();
        let InvalidProperty { property, message } = result_err;
        assert_eq!(property, "jump_by_pixel_min");
        assert_eq!(message, "jump_by_pixel_min cannot be equal to zero");

        // ----------------

        let config = Config {
            stayawake_interval: default_stayawake_interval(),
            jump_by_pixel_min: 1,
            jump_by_pixel_max: 0,
            init_point: InitPoint { x: 0, y: 0 },
            working_area: WorkingArea {
                width: 150,
                height: 150,
            },
        };

        let result = config.validate();
        assert!(result.is_err());

        let result_err = result.unwrap_err();
        let InvalidProperty { property, message } = result_err;
        assert_eq!(property, "jump_by_pixel_max");
        assert_eq!(message, "jump_by_pixel_max cannot be equal to zero");

        // ----------------

        let config = Config {
            stayawake_interval: default_stayawake_interval(),
            jump_by_pixel_min: 100,
            jump_by_pixel_max: 100,
            init_point: InitPoint { x: 0, y: 0 },
            working_area: WorkingArea {
                width: 0,
                height: 0,
            },
        };

        let result = config.validate();
        assert!(result.is_err());

        let result_err = result.unwrap_err();
        let InvalidProperty { property, message } = result_err;
        assert_eq!(property, "working_area");
        assert_eq!(
            message,
            "working_area height or/and width cannot be equal to zero"
        );

        // ----------------

        let config = Config {
            stayawake_interval: default_stayawake_interval(),
            jump_by_pixel_min: 100,
            jump_by_pixel_max: 100,
            init_point: InitPoint { x: 0, y: 0 },
            working_area: WorkingArea {
                width: 0,
                height: 150,
            },
        };

        let result = config.validate();
        assert!(result.is_err());

        let result_err = result.unwrap_err();
        let InvalidProperty { property, message } = result_err;
        assert_eq!(property, "working_area");
        assert_eq!(
            message,
            "working_area height or/and width cannot be equal to zero"
        );
    }
}
